compact.go

```diff
@@ -835,12 +846,55 @@ func (c *LeveledCompactor) populateBlock(blocks []BlockReader, meta *BlockMeta,
 			return errors.Wrap(err, "write label index")
 		}
 	}
+	names = append(names, apkName)
+	sort.Strings(names)
 
-	for _, l := range postings.SortedKeys() {
-		if err := indexw.WritePostings(l.Name, l.Value, postings.Get(l.Name, l.Value)); err != nil {
-			return errors.Wrap(err, "write postings")
+	if idxw, ok := indexw.(*index.Writer); ok {
+		idxw.HintPostingsWriteCount(numLabelValues)
+	}
+
+	postingBuf := make([]uint64, 0, 1000000)
+	var bigEndianPost index.Postings = index.NewBigEndianPostings(nil)
+	var listPost = index.NewListPostings(nil).(*index.ListPostings)
+	for _, n := range names {
+		s = s[:0]
+		if n == apkName {
+			s = append(s, apkValue)
+		} else {
+			for v := range values[n] {
+				s = append(s, v)
+			}
+			sort.Strings(s)
+		}
+
+		for _, v := range s {
+			postingBuf = postingBuf[:0]
+			for i, ir := range indexReaders {
+				bigEndianPost, err = ir.Postings(n, v, bigEndianPost)
+				if err != nil {
+					return errors.Wrap(err, "read postings")
+				}
+				for bigEndianPost.Next() {
+					if newVal, ok := set.seriesMap[i][bigEndianPost.At()]; ok {
+						// idx is the index at which newVal exists or index at which we need to insert.
```

Comment:
We could reverse the case here and handle `!ok` first to reduce indent?

Reply:
Done